tempodb/wal/append_block.go

```diff
@@ -52,6 +60,78 @@ func newAppendBlock(id uuid.UUID, tenantID string, filepath string, e backend.En
 	return h, nil
 }
 
+// newAppendBlockFromFile returns an AppendBlock that can not be appended to, but can
+// be completed. It is meant for wal replay
+func newAppendBlockFromFile(filename string, path string) (*AppendBlock, error) { // jpe test
+	blockID, tenantID, version, e, err := parseFilename(filename)
+	if err != nil {
+		return nil, err
+	}
+
+	v, err := encoding.FromVersion(version)
+	if err != nil {
+		return nil, err
+	}
+
+	b := &AppendBlock{
+		meta:     backend.NewBlockMeta(tenantID, blockID, version, e),
+		filepath: path,
+		encoding: v,
+	}
+
+	// replay file to extract records
+	f, err := b.file()
+	if err != nil {
+		return nil, err
+	}
+
+	dataReader, err := b.encoding.NewDataReader(backend.NewContextReaderWithAllReader(f), b.meta.Encoding)
+	if err != nil {
+		return nil, err
+	}
+	defer dataReader.Close()
+
+	var buffer []byte
+	var records []*common.Record
+	objectReader := b.encoding.NewObjectReaderWriter()
+	currentOffset := uint64(0)
+	for {
+		buffer, pageLen, err := dataReader.NextPage(buffer)
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+
+		reader := bytes.NewReader(buffer)
+		id, _, err := objectReader.UnmarshalObjectFromReader(reader)
+		if err != nil {
+			return nil, err
```

Comment:
It looks like a proto unmarshal failure here will cause the entire WAL file to be skipped and deleted up the call chain in `AllBlocks1.  I.e. if the last record wasn't flushed properly then the whole block will be lost ?   Previously I think it would have preserved the records that it could.    Maybe breaking out of the loop and continuing would have that effect here.  Returning it as a warning up the call-chain too.

Reply:
Good catch. Partial replay support added + tests.